### DISCUSSION TITLE
Kepler GPU not advertising ReBAR capability on PCIe bus

### DISCUSSION THREAD

**Discussion by terminatorul** [terminatorul/NvStrapsReBar#18] [Q&A] [answered]
Hi @Xelafic, are you still around please ?

I tried to enable ReBAR on my very low end Kepler GPU (GK208, GeForce GT 710), and it looks like ReBAR is not advertised on the PCI 2.0 bus (see image, also CPU-Z report attached). This is true even after seting straps bits.

I rememeber the original discussion where you showed me a 2GB BAR on your GT 640 (Kepler GK107).

Can you tell me how did you get it working ? Did you do something special ? I see the straps bits should be the same ... did you see any differences ?

@xCuri0 do you know what can I do if ReBAR capability is not present in the PCI configuration space ?

[X399-AORUS-GAMI-GT710.txt](https://github.com/terminatorul/NvStrapsReBar/files/14266293/X399-AORUS-GAMI-GT710.txt)

![image](https://github.com/terminatorul/NvStrapsReBar/assets/378924/3f56d704-34a3-4cca-97cd-0747e2eedc25)

**Comment by xCuri0** [comment]
@terminatorul I think Xelafic said Pascal and older only expose the strap bits, the BAR size changes using them.

**Reply by xCuri0** [reply]
Being said there is the slim possibility of a strap bit that exposes resizable BAR but even if that was true it would most likely be on Pascal only.


Maybe the nouveau driver team would possibly know more about the straps than outdated envytools documentation

**Reply by terminatorul** [reply]
Looks like @Xelafic really got a Keppler card going, see the last screenshot in the message here: https://github.com/xCuri0/ReBarUEFI/discussions/89#discussioncomment-7596828

**Reply by xCuri0** [reply]
@terminatorul isn't that same as what you are doing now ? just changing strap bits and getting large BAR size

**Reply by terminatorul** [reply]
Now I am setting the straps bits, and the the GPU still has no ReBAR extended capability in the PCI configuration space

**Reply by xCuri0** [reply, marked as answer]
@terminatorul xelafic also did not get resizable bar you can see in the screenshot.

What you will find though is that device manager shows the GPU has increased BAR size. And if the driver working correctly with it the vulkan memory pool in GPU-Z will too

**Reply by terminatorul** [reply]
Oh, I see, this must explain the empty output in GPU-Z, in the screenshot from @Xelafic.

For my board it stil didn't work 'though, Device Manager says "This device is not using any resources because it has a problem" ..., and it shows the device with an exclamation mark !

:(

**Reply by xCuri0** [reply]
Will have to wait for Xelafics response then maybe Kepler needs to be configured a bit different.

**Reply by ghost** [reply]
Hi @terminatorul , you've been busy! 

Kepler uses offset BAR0 offset 0x101000 and bits 17:14 IINM to give a BAR1 range of 64MiB to 64GiB. AFAIK nvidia Geforce Resizable BAR capability started with Turing. Models before Turing can have BAR1 resized directly by their straps as there is no override via Resizable BAR capability. 

FWIW I also tested an early Maxwell GTX 750Ti and that behaves similar to Pascal using BAR0 offset 0x10100c for strap bits. 

BTW the claimed author of GPU-Z is asking for someone to test Turing Resizable BAR with a new GPU-Z build? See https://github.com/xCuri0/ReBarUEFI/discussions/118#discussioncomment-8442136 I think better it works with systems enabled by your software foremost. Or perhaps  @UnidentifiedTag can do it?

**Reply by terminatorul** [reply]
Nice to hear from you again :)

Looks like straps 0 bits 17:14 do not quite match the envytools page at https://envytools.readthedocs.io/en/latest/hw/io/pstraps.html, most likely you meant bits 17:16, with a max of 512 MiB size. Do you have some different information about this ?

I have a possible scenario now:

So straps bits for BAR1 size are grouped in 2 parts, that are added together:
- straps0 (0x101000) bits 17:16
- straps1 (0x10100C) bits 22:20

NvStrapsReBar driver uses different sets depending on the target BAR size:
- first set (straps 0 bits 17:16) is used for a BAR size up to 256MB
- mostly the second set is used for larger BAR sizes, and the first set is kept at the default of `10` (that originally indicates 256 MiB)

Now the GK208 chips works when setting BAR1 size from 128 to 256, and does not work for higher values.

So it indicates straps0 bits are working, and straps1 bits are triggering the error in the Device Manager. I will check this again when I get back to my Keppler GPU...

But my straps0 bits (17:16) only allow for 512 MiB BAR size max. Can you check if this is right please ? So how did you get 2 GiB then ?

(P.S: Turin**g** - with a **g**, instead of Turin...)

**Reply by terminatorul** [reply]
P.S.  Maybe Pascal has the same problem, too, so I could set a 512 MiB BAR for Pascal !!!!

**Reply by terminatorul** [reply]
After testing still can't use 512 MiB or more... My GK208 is like a Pascal card, nothing above 256 MiB works, except I get a device error in Device Manager instead of the BSOD for Pascal.

**Reply by ghost** [reply]
> (P.S: Turin**g** - with a **g**, instead of Turin...)

Wow, didn't even realize it after all this time. Too funny to be sad, hope it wasn't too irritating. Thanks for the correction.

BAR1 tests on GT640 using offset 0x101000 and bits 17:14

```
BAR0 Reg	Bits	 BAR1
??101000	17:14	 Size
================================
8040049e	0000	 64MiB
8040449e	0001	128MiB
8040849e	0010	256MiB    Default
8040c49e	0011	512MiB
8041049e	0100	  1GiB
8041449e	0101	  2GiB
8041849e	0110	  4GiB
8041c49e	0111	  8GiB
8042049e	1000	 16GiB
8042449e	1001	 32GiB
8042849e	1010	 64GiB
8042c49e	1011	128GiB
8043049e	1100	256GiB
8043449e	1101	512GiB
8043849e	1110	  1TiB
8043c49e	1111	  1TiB
```

Not sure why the last 2 are the same, maybe a limit? Still, should be more than enough for the 2GiB of VRAM :grinning:

**Reply by xCuri0** [reply]
@Xelafic how did you figure out it's 17:14 ? It's not documented anywhere


Can you show what the GPU-Z vulkan tab memory looks like after doing this with 2GB ? I want to see if the driver using it

**Reply by ghost** [reply]
By setting bits. 

There's also bits 22:20 at offset 0x10100c which would give up to 8GiB if wanting to use those or both can be used together where adding the bits and using them as a power of 2 * 64MiB to give a theoretical max of 128TiB IINM. I can only test to 32TiB though due to CPU physical address space being 46-bit.

Re Vulkan, I'm not set up for it at this time but this card is not supported by nvidia-smi for BAR1 checking and therefore probably not supported for Vulkan. There is another test I want to do with this card so will try it then.

Edit:
![vulk](https://github.com/terminatorul/NvStrapsReBar/assets/147298897/8b31e884-c936-4e8d-ac7e-4816044e0eb1)

**Comment by terminatorul** [comment]
Looks like the stock BAR size for this card is 128 GiB instead of the usual 256 GiB.

I can set up to 2 GiB BAR with Microsoft Basic Display Adapter, and I can set 256 GiB (instead of 128 GiB) with the NVIDIA driver 474.64

Maybe my GT 710 card is just too low end for large BAR sizes :(

**Reply by xCuri0** [reply]
I think you meant MiB ?

**Reply by terminatorul** [reply]
Yes, indeed

**Reply by ghost** [reply]
Does it show 128MiB under Resizable BAR on GPU-Z? Suggested last year it would be nice to see BAR sizes for all cards instead of 'unsupported GPU'. They didn't seem to be aware of cards existing with sizes other than 256MiB! Don't know if something was done or not.

**Reply by terminatorul** [reply]
No, I used the output of `NvStrapsReBar.exe` to see the 128MB BAR. `NvStrapsReBar.exe` enumerates the memory resources from Device Manager, and picks the larger memory size as the current GPU BAR size.

I also sugested to the GPU-Z maintainer / developer to show a large BAR size even if it is not resizable. The answer was the API used by GPU-Z, to read the BAR size, no longer works if the BAR is not resizable (so without ReBAR enabled). GPU-Z is closed source, so I don't know what API it uses to read PCI BAR size.

**Comment by terminatorul** [comment]
> BAR1 tests on GT640 using offset 0x101000 and bits 17:14
> 
> ```
> BAR0 Reg	Bits	 BAR1
> ??101000	17:14	 Size
> ================================
> 8040049e	0000	 64MiB
> 8040449e	0001	128MiB
> 8040849e	0010	256MiB    Default
> 8040c49e	0011	512MiB
> 8041049e	0100	  1GiB
> 8041449e	0101	  2GiB
> 8041849e	0110	  4GiB
> 8041c49e	0111	  8GiB
> 8042049e	1000	 16GiB
> 8042449e	1001	 32GiB
> 8042849e	1010	 64GiB
> 8042c49e	1011	128GiB
> 8043049e	1100	256GiB
> 8043449e	1101	512GiB
> 8043849e	1110	  1TiB
> 8043c49e	1111	  1TiB
> ```
> 
> Not sure why the last 2 are the same, maybe a limit? Still, should be more than enough for the 2GiB of VRAM 😀

@Xelafic  have you considered reporting these findings to [envytools](https://github.com/envytools/envytools) ?

**Reply by ghost** [reply]
Given the straps data hasn't been updated in years on envytools I believe the team has moved on to other agendas. Probably they are already aware but just don't have the resources and time to add it. Given that, it would seem inappropriate to mention it. Saw your post on VRAM from BAR0 BTW. Thanks, would never have found it myself with such a convoluted method!

Since we are talking about straps, FWIW I have seen an Nvidia post showing a Quadro K600 with nvidia-smi memory command supported. The straps on the GT640 also allow changing the device ID so I configured the GT640 to report as a K600 which allowed installing the Quadro drivers which worked well with the card but still didn't produce nvidia-smi data for memory :disappointed: Perhaps something needed in the vbios? I didn't try flashing the K600 vbios to the GT640 as it seemed too incompatible.
